fuel-vm/src/interpreter/receipts.rs

```diff
@@ -0,0 +1,170 @@
+use core::ops::Index;
+use std::mem;
+
+use fuel_merkle::binary;
+use fuel_tx::Receipt;
+use fuel_types::{bytes::SerializableVec, Bytes32};
+
+#[derive(Debug, Default, Clone)]
+pub(crate) struct ReceiptsCtx {
+    receipts: Vec<Receipt>,
+    receipts_tree: binary::in_memory::MerkleTree,
+}
+
+impl ReceiptsCtx {
+    pub fn push(&mut self, mut receipt: Receipt) {
+        self.receipts_tree.push(receipt.to_bytes().as_slice());
+        self.receipts.push(receipt)
+    }
+
+    pub fn clear(&mut self) {
+        self.receipts_tree.reset();
+        self.receipts.clear();
+    }
+
+    pub fn root(&self) -> Bytes32 {
+        self.receipts_tree.root().into()
+    }
+
+    /// Get a mutable lock on this context
+    pub fn lock(&mut self) -> ReceiptsCtxMut {
```

Comment:
Tried to think of another way this could be done, but this seems to strike the best balance between minimizing allocs and safety when rewinding diffs.